src/uri_utils.cpp

```diff
@@ -0,0 +1,454 @@
+
+inline bool is_userinfo_char(char c)
+{
+    return is_unreserved(c) || is_subdelim(c) || c == ':' || c == '%';
+}
+
+inline bool is_regname_char(char c) { return is_unreserved(c) || is_subdelim(c) || c == '%'; }
+
+} // namespace
+
+std::optional<uri_decomposed> uri_parse(std::string_view uri)
+{
+    uri_decomposed decomposed;
+    decomposed.raw = uri;
+
+    auto expected_token = token_type::scheme;
+    auto lookahead_token = token_type::none;
+
+    // Authority helpers
+    std::size_t authority_end = npos;
+    std::string_view authority_substr;
+
+    for (std::size_t i = 0; i < uri.size();) {
+        // Dead man's switch
+        auto current_token = expected_token;
+        expected_token = token_type::none;
+
+        switch (current_token) {
+        case token_type::scheme: {
+            auto token_begin = i;
+            if (!isalpha(uri[i++])) {
+                // The URI is malformed as the first character must be alphabetic
+                return std::nullopt;
+            }
+
+            bool end_found = false;
+            while (i < uri.size()) {
+                const auto c = uri[i++];
+                if (c == ':') {
+                    // We reached the end of the scheme, move to the next token
+                    end_found = true;
+                    break;
+                }
+
+                if (!is_scheme_char(c)) {
+                    return std::nullopt;
+                }
+            }
+
+            if (!end_found) {
+                return std::nullopt;
+            }
+
+            expected_token = token_type::hierarchical_part;
+            decomposed.scheme = uri.substr(token_begin, i - 1);
+
+            break;
+        }
+        case token_type::hierarchical_part: {
+            if ((i + 1) < uri.size() && uri[i] == '/' && uri[i + 1] == '/') {
+                // The authority always starts with //
+                expected_token = token_type::authority;
+                i += 2;
+            } else {
+                // Otherwise we expect a path (path-absolute, path-rootless, path-empty)
+                expected_token = token_type::path_no_authority;
+            }
+            break;
+        }
+        case token_type::path_no_authority: {
+            auto token_begin = i;
+            // The path can be empty but we wouldn't be here...
+            while (i < uri.size()) {
+                const auto c = uri[i++];
+                if (!is_path_char(c) && c != '/') {
+                    return std::nullopt;
+                }
+            }
+
+            decomposed.path_index = token_begin;
+            decomposed.path = uri.substr(token_begin, i - token_begin);
+
+            // We're done, nothing else to parse
+            return decomposed;
+        }
+        case token_type::authority: {
+            auto token_begin = i;
+            authority_end = uri.find_first_of("/?#", i);
+            if (authority_end != npos) {
+                // The authority is empty
```

Comment:
I don't get this comment. This would only be the case if `authority_end == i`, no?

Reply:
Ah no, I had structured the code differently before and forgot to remove it.